packages/dds/tree/src/feature-libraries/sequence-change-family/compose.ts

```diff
@@ -0,0 +1,209 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+    getMarkLength,
+    isAttachGroup,
+    isDetachMark,
+    isReattach,
+    MarkListFactory,
+    splitMark,
+    Transposed as T,
+} from "../../changeset";
+import { clone, fail } from "../../util";
+import { SequenceChangeset } from "./sequenceChangeset";
+
+/**
+ * Composes a sequence of changesets into a single changeset.
+ * @param changes - The changesets to be applied.
+ * Parts of the input may be reused in the output, but the input is not mutated.
+ * Each changeset in the list is assumed to be applicable after the previous one.
+ * @returns A changeset that is equivalent to applying each of the given `changes` in order.
+ *
+ * WARNING! This implementation is incomplete:
+ * - Tombstone information is ignored.
+ * - Support for moves is not implemented.
+ * - Support for slices is not implemented.
+ */
+export function compose(changes: SequenceChangeset[]): SequenceChangeset {
+    if (changes.length === 1) {
+        return changes[0];
+    }
+    const base: SequenceChangeset = {
+        marks: {},
+    };
+    for (const change of changes) {
+        foldInFieldMarks(change.marks, base.marks);
+    }
+    return base;
+}
+
+function foldInFieldMarks(newFieldMarks: T.FieldMarks, baseFieldMarks: T.FieldMarks) {
+    for (const key of Object.keys(newFieldMarks)) {
+        const newMarkList = newFieldMarks[key];
+        baseFieldMarks[key] ??= [];
+        const baseMarkList = foldInMarkList(newMarkList, baseFieldMarks[key]);
+        baseFieldMarks[key] = baseMarkList;
+        while (typeof baseMarkList[baseMarkList.length - 1] === "number") {
+            baseMarkList.pop();
+        }
+        if (baseMarkList.length === 0) {
+            // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+            delete baseFieldMarks[key];
+        }
```

Comment:
This would fail to delete `baseFieldMarks[key]` in cases where it starts populated but needs to be deleted. I agree this code can be simplified though. I've given it a go.